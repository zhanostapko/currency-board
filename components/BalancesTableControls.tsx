import Button from "./Shared/Button";
import Input from "./Shared/Input";

type BalancesTableControlsProps = {
  searchQuery: string;
  onSearchChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  columnCount: number;
  maxColumns: number;
  increment: () => void;
  decrement: () => void;
  error?: string | null;
};

const BalancesTableControls = ({
  searchQuery,
  onSearchChange,
  columnCount,
  maxColumns,
  increment,
  decrement,
  error,
}: BalancesTableControlsProps) => {
  return (
    <div className="flex flex-wrap items-center justify-center gap-4 mb-6">
      <Input
        disabled={!!error}
        value={searchQuery}
        onChange={onSearchChange}
        placeholder="Search"
        className="w-64 px-4 py-2 border border-gray-300 rounded-md bg-white text-gray-700 placeholder-gray-500 shadow-sm focus:ring-2 focus:ring-blue-500 focus:border-blue-500 transition disabled:bg-gray-100 disabled:cursor-not-allowed disabled:hover:bg-gray-100"
      />
      <div className="flex gap-4">
        <Button
          className="px-5 py-2.5 text-sm bg-blue-600 text-white rounded-md hover:bg-blue-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
          disabled={columnCount >= maxColumns || !!error}
          onClick={increment}
        >
          + Increase
        </Button>
        <Button
          className="px-5 py-2.5 text-sm bg-red-600 text-white rounded-md hover:bg-red-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
          disabled={columnCount <= 1 || !!error}
          onClick={decrement}
        >
          - Reduce
        </Button>
      </div>
    </div>
  );
};

export default BalancesTableControls;
