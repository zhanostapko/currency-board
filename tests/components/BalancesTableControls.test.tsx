import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import BalancesTableControls from "../../components/BalancesTableControls";

describe("BalancesTableControls Component", () => {
  const mockOnSearchChange = jest.fn();
  const mockIncrement = jest.fn();
  const mockDecrement = jest.fn();

  const defaultProps = {
    searchQuery: "",
    onSearchChange: mockOnSearchChange,
    columnCount: 2,
    maxColumns: 5,
    increment: mockIncrement,
    decrement: mockDecrement,
    error: null,
  };

  test("renders input and buttons", () => {
    const { asFragment } = render(<BalancesTableControls {...defaultProps} />);
    expect(screen.getByPlaceholderText("Search")).toBeInTheDocument();
    expect(screen.getByText("+ Increase")).toBeInTheDocument();
    expect(screen.getByText("- Reduce")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test("calls onSearchChange when input is changed", () => {
    render(<BalancesTableControls {...defaultProps} />);
    const input = screen.getByPlaceholderText("Search");

    fireEvent.change(input, { target: { value: "BTC" } });

    expect(mockOnSearchChange).toHaveBeenCalledTimes(1);
  });

  test("calls increment function when + button is clicked", () => {
    render(<BalancesTableControls {...defaultProps} />);
    const increaseButton = screen.getByText("+ Increase");

    fireEvent.click(increaseButton);

    expect(mockIncrement).toHaveBeenCalledTimes(1);
  });

  test("calls decrement function when - button is clicked", () => {
    render(<BalancesTableControls {...defaultProps} />);
    const decreaseButton = screen.getByText("- Reduce");

    fireEvent.click(decreaseButton);

    expect(mockDecrement).toHaveBeenCalledTimes(1);
  });

  test("disables buttons when columnCount reaches limits", async () => {
    const { asFragment } = render(
      <BalancesTableControls {...defaultProps} columnCount={5} maxColumns={5} />
    );
    expect(screen.getByRole("button", { name: "+ Increase" })).toBeDisabled();

    render(<BalancesTableControls {...defaultProps} columnCount={1} />);

    await waitFor(() => {
      const reduceButton = screen
        .getAllByRole("button")
        .find((btn) => (btn as HTMLButtonElement).disabled);
      expect(reduceButton).toBeDisabled();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  test("disables input and buttons when error exists", () => {
    const { asFragment } = render(
      <BalancesTableControls {...defaultProps} error="Test Error" />
    );
    expect(screen.getByPlaceholderText("Search")).toBeDisabled();
    expect(screen.getByText("+ Increase")).toBeDisabled();
    expect(screen.getByText("- Reduce")).toBeDisabled();
    expect(asFragment()).toMatchSnapshot();
  });
});
