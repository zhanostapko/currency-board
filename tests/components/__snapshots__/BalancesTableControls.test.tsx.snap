// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BalancesTableControls Component disables buttons when columnCount reaches limits 1`] = `
<DocumentFragment>
  <div
    class="flex flex-wrap items-center justify-center gap-4 mb-6"
  >
    <input
      class="w-64 px-4 py-2 border border-gray-300 rounded-md bg-white text-gray-700 placeholder-gray-500 shadow-sm focus:ring-2 focus:ring-blue-500 focus:border-blue-500 transition disabled:bg-gray-100 disabled:cursor-not-allowed disabled:hover:bg-gray-100"
      placeholder="Search"
      value=""
    />
    <div
      class="flex gap-4"
    >
      <button
        class="px-5 py-2.5 text-sm bg-blue-600 text-white rounded-md hover:bg-blue-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
        disabled=""
      >
        + Increase
      </button>
      <button
        class="px-5 py-2.5 text-sm bg-red-600 text-white rounded-md hover:bg-red-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
      >
        - Reduce
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BalancesTableControls Component disables input and buttons when error exists 1`] = `
<DocumentFragment>
  <div
    class="flex flex-wrap items-center justify-center gap-4 mb-6"
  >
    <input
      class="w-64 px-4 py-2 border border-gray-300 rounded-md bg-white text-gray-700 placeholder-gray-500 shadow-sm focus:ring-2 focus:ring-blue-500 focus:border-blue-500 transition disabled:bg-gray-100 disabled:cursor-not-allowed disabled:hover:bg-gray-100"
      disabled=""
      placeholder="Search"
      value=""
    />
    <div
      class="flex gap-4"
    >
      <button
        class="px-5 py-2.5 text-sm bg-blue-600 text-white rounded-md hover:bg-blue-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
        disabled=""
      >
        + Increase
      </button>
      <button
        class="px-5 py-2.5 text-sm bg-red-600 text-white rounded-md hover:bg-red-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
        disabled=""
      >
        - Reduce
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`BalancesTableControls Component renders input and buttons 1`] = `
<DocumentFragment>
  <div
    class="flex flex-wrap items-center justify-center gap-4 mb-6"
  >
    <input
      class="w-64 px-4 py-2 border border-gray-300 rounded-md bg-white text-gray-700 placeholder-gray-500 shadow-sm focus:ring-2 focus:ring-blue-500 focus:border-blue-500 transition disabled:bg-gray-100 disabled:cursor-not-allowed disabled:hover:bg-gray-100"
      placeholder="Search"
      value=""
    />
    <div
      class="flex gap-4"
    >
      <button
        class="px-5 py-2.5 text-sm bg-blue-600 text-white rounded-md hover:bg-blue-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
      >
        + Increase
      </button>
      <button
        class="px-5 py-2.5 text-sm bg-red-600 text-white rounded-md hover:bg-red-700 transition flex items-center gap-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
      >
        - Reduce
      </button>
    </div>
  </div>
</DocumentFragment>
`;
